evm/vm/evm.py

```diff
@@ -243,30 +236,42 @@ def import_block(self, block):
         except BlockNotFound:
             raise ValidationError("Parent ({0}) of block {1} not found".format(
                 block.header.parent_hash, block.header.hash))
+
         init_header = self.create_header_from_parent(parent_header)
         parent_evm = type(self)(self.db, init_header)
         imported_block = parent_evm.get_vm().import_block(block)
+        # It feels wrong to call validate_block() on self here, but we do that
+        # because we want to look up the recent uncles starting from the
+        # currenct canonical chain head.
         self.validate_block(imported_block)
 
         persist_block_to_db(self.db, imported_block)
-        # TODO: We must only do this when the imported block has higher
-        # score than current head.
-        self.header = self.create_header_from_parent(imported_block.header)
+        self.maybe_add_to_canonical_chain(imported_block)
 
         return imported_block
 
-    def mine_block(self, **mine_params):
-        """
-        Mine the current block, applying
-        """
-        block = self.get_vm().mine_block(**mine_params)
-        self.validate_block(block)
-        persist_block_to_db(self.db, block)
-
+    def maybe_add_to_canonical_chain(self, block):
```

Comment:
Something doesn't sit well with me regarding the *maybe* aspect of this method.  I'm thinking it may be better to split this into two pieces.

* a function that returns whether the block should be the new canonical head of the chain.
* a function that sets the block as the new canonical head of the chain (regardless of the score).

Then we perform the check, and if the check confirms it should be the new head block, we then set it as the head block.  Thoughts?